### PR TITLE
PP-3510 - adding friendly url to the flow 

### PR DESCRIPTION
**Pull request by jonheslop**
Only merge after EndToEnd: https://github.com/alphagov/pay-endtoend/pull/586

## 688433f - adding slugify package to format urls
This is a tiny package with no other dependencies that makes it easy for
us to consistently format urls client and server side.

## c36bdc3 - Adding url playback confirmation 
- slugify is accessible as a nunjucks filter for ease of use
- extending input confirm to be able to use slugify
- also removing indefinite article (a/an) from the url to keep the url
short.
- remove the definite article (the) too

## f6c4540 - Passed friendly URL to link creation 
- Added friendly URL to the session
- extended products client to include friendly URL bits
- prefer display of friendly URL on manage page

### NOTE: 
This also contains a backward compatible fix to make the JS work in PhantomJS

### CODE REVIEW

File: app/controllers/payment-links/post-information-controller.js [hunk 1]
```diff
@@ -12,8 +13,12 @@ module.exports = (req, res) => {
 
   updatedPageData.paymentLinkTitle = req.body['payment-link-title']
   updatedPageData.paymentLinkDescription = req.body['payment-link-description']
+  updatedPageData.serviceNamePath = req.body['service-name-path']
+  updatedPageData.productNamePath = slugify(req.body['payment-link-title'].replace(/\ba\s|\ban\s|\bthe\b/gi, ''))
   lodash.set(req, 'session.pageData.createPaymentLink', updatedPageData)
 
+  console.log(updatedPageData)
```
Comment (ckalista) [inline app/controllers/payment-links/post-information-controller.js]: To remove.

File: app/controllers/payment-links/post-information-controller.js [hunk 2]
```diff
@@ -12,8 +13,12 @@ module.exports = (req, res) => {
 
   updatedPageData.paymentLinkTitle = req.body['payment-link-title']
   updatedPageData.paymentLinkDescription = req.body['payment-link-description']
+  updatedPageData.serviceNamePath = req.body['service-name-path']
```
Comment (ckalista) [inline app/controllers/payment-links/post-information-controller.js]: What have we decided for the service name. I reckon the same normalisation rules apply as for product name. May we should create a separate function that we can reuse as we type the title and then again when we validate the input fields after an update?

File: app/controllers/payment-links/post-information-controller.js [hunk 3]
```diff
@@ -12,8 +13,12 @@ module.exports = (req, res) => {
...
+  updatedPageData.productNamePath = slugify(req.body['payment-link-title'].replace(/\ba\s|\ban\s|\bthe\b/gi, ''))
```
Comment (ckalista) [inline app/controllers/payment-links/post-information-controller.js]: We need to have test coverage for this.

File: app/controllers/payment-links/post-review-controller.js [hunk 4]
```diff
@@ -54,7 +56,9 @@ module.exports = (req, res) => {
         paymentLinkTitle,
         paymentLinkDescription,
         paymentLinkAmount,
-        req.service.name
+        req.service.name,
+        serviceNamePath,
+        productNamePath
```
Comment (ckalista) [inline app/controllers/payment-links/post-review-controller.js]: We need test coverage for this.

Comment (jonheslop) [reply app/controllers/payment-links/post-review-controller.js]: Added tests for everything

File: app/controllers/payment-links/get-web-address-controller.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const lodash = require('lodash')
+
+// Local dependencies
+const {response} = require('../../utils/response.js')
+const paths = require('../../paths')
+
+module.exports = (req, res) => {
+  const friendlyURL = process.env.PRODUCTS_FRIENDLY_BASE_URI
+  const pageData = lodash.get(req, 'session.pageData.createPaymentLink', {})
+  const productNamePath = req.body['payment-amount'] || pageData.productNamePath || ''
```
[the remark this answers is not in the document]
Comment (ckalista) [reply app/controllers/payment-links/get-web-address-controller.js]: Good spot, I believe `req.body['payment-amount']` is always going to be undefined hence can safely be removed.

File: app/browsered.js [hunk 6]
```diff
@@ -10,6 +10,8 @@ const fieldValidation = require('./browsered/field-validation')
 const dashboardActivity = require('./browsered/dashboard-activity')
 const targetToShow = require('./browsered/target-to-show')
 const analytics = require('gaap-analytics')
+const inputConfirm = require('./browsered/input-confirm')
+const niceURL = require('./browsered/nice-url')
```
[the remark this answers is not in the document]
Comment (ckalista) [reply app/browsered.js]: Technically this is not the friendly URL we are building here so I will leave it as it is.

**Pull request merged**